src/lib/consensus/proof_of_stake.ml

```diff
@@ -2876,69 +2691,82 @@ module Hooks = struct
               } ;
             Deferred.Or_error.ok_unit )
       else
-        let%bind peers = random_peers 5 in
-        Deferred.List.fold peers
-          ~init:(Or_error.error_string "Failed to sync epoch ledger: No peers")
-          ~f:(fun acc peer ->
-            match acc with
+        let num_sync_peers = 5 in
+        let%bind peers = random_peers num_sync_peers in
+        let ledger_hash_json =
+          Mina_base.Ledger_hash.to_yojson target_ledger_hash
+        in
+        [%log info] "Syncing epoch ledger from %d peers" num_sync_peers
+          ~metadata:[ ("target_ledger_hash", ledger_hash_json) ] ;
+        (* start with an existing epoch ledger, which may be faster
+           than syncing with an empty ledger, since ledgers accumulate
+           new leaves in increasing index order
+        *)
+        let%bind.Deferred.Or_error db_ledger =
+          let db_ledger_of_snapshot snapshot =
+            match snapshot.ledger with
+            | Ledger_snapshot.Ledger_db ledger ->
+                Ok ledger
+            | Ledger_snapshot.Genesis_epoch_ledger ledger ->
+                let module Ledger_transfer =
+                  Mina_ledger.Ledger_transfer.Make
+                    (Mina_ledger.Ledger)
+                    (Mina_ledger.Ledger.Db)
+                in
+                let fresh_db_ledger =
+                  Mina_ledger.Ledger.Db.create
+                    ~depth:Context.constraint_constants.ledger_depth ()
+                in
+                Ledger_transfer.transfer_accounts ~src:ledger
+                  ~dest:fresh_db_ledger
+          in
+          match snapshot_id with
+          | Staking_epoch_snapshot ->
+              return
+              @@ db_ledger_of_snapshot !local_state.staking_epoch_snapshot
+          | Next_epoch_snapshot ->
+              return @@ db_ledger_of_snapshot !local_state.next_epoch_snapshot
+        in
+        let sync_ledger =
+          Mina_ledger.Sync_ledger.Db.create ~logger ~trust_system db_ledger
+        in
+        let query_reader =
+          Mina_ledger.Sync_ledger.Db.query_reader sync_ledger
+        in
+        let response_writer =
+          Mina_ledger.Sync_ledger.Db.answer_writer sync_ledger
+        in
+        glue_sync_ledger ~preferred:peers query_reader response_writer ;
```

Comment:
The preferred set should probably start either as empty or as peers that we interacted with successfully when syncing the root snarked ledger. It seems wrong to set this set to a random subset of 5 peers when we have no reason to believe those 5 peers have the epoch ledgers we need or not.

Reply:
That was the strategy that had been used, right?

Reply:
The strategy that is currently being used, according to the code, is to pass the peer that sent us the current best tip, or to specify no preferred peers at all https://github.com/MinaProtocol/mina/blob/develop/src/lib/bootstrap_controller/bootstrap_controller.ml#L300.

Reply:
Sorry, I meant the current RPC to download epoch ledgers seems to choose 5 random peers: https://github.com/MinaProtocol/mina/blob/develop/src/lib/consensus/proof_of_stake.ml#L2955